### ISSUE TITLE
sub plugins are not being pulled into documentation or ansible-doc etc

### ISSUE THREAD

**Issue by samccann** [ansible/ansible#75109]
### Summary

Some collections, such as ansible.utils, have an entity called sub_plugins that include DOCUMENTATION blocks that are not visible in docs.ansible.com or probably not in `ansible-doc`.  These sub_plugins are important to end users and their information should be displayed.

Examples of sub_plugins:

https://github.com/ansible-collections/ansible.utils/tree/main/plugins/sub_plugins

### Issue Type

Documentation Report

### Component Name

docs/docsite/rst/

### Ansible Version

```console
2.9, 2.11
```

### Configuration

```console
$ ansible-config dump --only-changed
```

### OS / Environment

Fedora 33

### Additional Information

This added information will improve the user experience.

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by sivel** [comment]
`sub_plugins` isn't actually a valid plugin type.  It's just a reserved directory that collections are permitted to utilize, but they have no real meaning in the context of our plugins system.

**Comment by samccann** [comment]
Thanks @sivel - does this mean we don't feel the DOCUMENTATION blocks within those `sub_plugins` have any valid home in our documentation?  I was specifically dealing with this one - https://github.com/ansible-collections/ansible.utils/blob/main/plugins/sub_plugins/validate/jsonschema.py  where it talks about options and even environment variables that a user wouldn't know about if we don't surface the docs somehow...

**Comment by samccann** [comment]
btw it's possible this was already discussed and decided that no we wouldn't' do anything to surface those docs....I'm just trying to see if that decision was already made, and if not - we should decide one way or the other.

**Comment by sivel** [comment]
Yeah, I'm not entirely sure.  Since they aren't a valid plugin type, `ansible-doc` doesn't know about them, their location just gives a standard place to store this type of code, that is importable by the collection, and in a directory that ansible-core will not eventually use.

I don't know what assumptions were made about future documentation capabilities, if any at all.

**Comment by felixfontein** [comment]
Right now the `sub_plugins` folder can be used by collections for random things. The convention of having something in there which has `DOCUMENTATION` is a convention invented and used by only very few collections (AFAIK). I think that these should be standardized in some way (maybe https://github.com/ansible/proposals/?) before they should be properly supported by core, to avoid more chaos.

**Comment by cidrblock** [comment]
Some additional information, in an effort to clarify a few things:
- Both ansible.utils and ansible.netcommon leverage a subplugin architecture and are fully supported with a Red Hat Ansible Automation Platform subscription.
- The subplugin architecture was introduced to provide a consistent playbook structure, while exposing 'engines' that leverage different python packages, and will work across collection boundaries
- The subplugin architecture requires no additional code changes in ansible-core for automation execution (since 2.9)
- The subplugin architecture uses a standards-based python import for subplugin loading and does not leverage ansible-core's "PluginLoader"
- Subplugins are individually documented and can be exposed, leveraged, and invoked by all other plugin types
- Typically the parent plugin introduces a capability, while the subplugin corresponds to the technical implementation
- The 'subplugins' directory is the agreed upon standard location for code supporting the subplugin architecture as of ansible 3.0
- ansible-navigator and collection-prep will continue to render documentation for all plugin types, including subplugins.

IMO, rendering documentation for all plugin types, including subplugins would be a win for ansible users.

**Comment by ganeshrn** [comment]
Raised propoal https://github.com/ansible/proposals/issues/194 to add support for sub-plugins within ansible-core

**Comment by jimi-c** [comment]
Because sub_plugins can be a catch-all for multiple types of plugins, we should not encourage its use. In fact it would be a major problem, as there could be more than one type of otherwise valid plugin in that directory. Instead, we should work on removing its use from our collections by relocating code there to module_utils or wherever we feel is appropriate.

**Comment by ganeshrn** [comment]
> In fact it would be a major problem, as there could be more than one type of otherwise valid plugin in that directory. 

If ansible-core standardize on the sub-plugin directory structure within the collection, in that case, the files within a particular folder can be identified as sub-plugins associated with a single top-level standard plugin avoiding the issue of having a valid top-level plugin in that directory. 

>. Instead, we should work on removing its use from our collections by relocating code there to module_utils or wherever we feel is appropriate.

Currently, code that doesn't fit in `module_utils` can be added in the `plugin_utils` directory within the collection but the `plugin_utils` folder doesn't solve the issue described in the proposal.

**Comment by felixfontein** [comment]
> The 'subplugins' directory is the agreed upon standard location for code supporting the subplugin architecture as of ansible 3.0

I would say that the `plugins/sub_plugins` directory is the agreed upon directory that can be used by collections for subplugins, but there is no requirement that these subplugins satisfy the additional requirements listed (like "Subplugins are individually documented"). See https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#modules--plugins for the precise specification.

**Comment by nitzmahone** [comment]
This will probably end up being covered by the planned docs support for additional collection resources that are unknown to core (eg, EDA plugins)- likely Ansible Core 2.18.

**Comment by felixfontein** [comment]
@nitzmahone is there any information on how this might/could look like?

**Comment by LeonxLJX** [comment]
I'd like to take this one.

Reading the issue, this looks like a documentation-clarity gap rather than a code defect: the current page doesn't give readers enough to self-diagnose, which is why it surfaces here as an issue instead of being answered by the docs.

My plan: locate the relevant section, confirm the described behavior against the current source (so the docs match what the code actually does today, not what it did when the page was written), then tighten the wording and add the missing piece. I'll also scan neighboring pages for the same inconsistency, otherwise the fix just relocates the confusion.

One thing I'd like to confirm before writing: would you prefer a minimal correction scoped to this page, or should I also cross-link it from the related guides if the same gap shows up there?

(claiming via @LeonxLJX)

**Comment by sivel** [comment]
I'm closing this out. There are no plans for ansible-doc to support sub plugins. ansible-doc will continue to support only plugins that core itself knows about and declares as an actual plugin type.

There is planned work elsewhere to create tooling to enumerate and extract documentation from non-core plugin types.

If you have further questions please stop by our forum or Matrix:

   * Forum: https://forum.ansible.com/
   * Matrix: https://matrix.to/#/#users:ansible.com
